### PR TITLE
Support downstream entries/spiffeIds (closes #43)

### LINKED ISSUE #43: Support downstream entries/spiffeIds
The Registrar supported creating entries/SpiffeIds with the downstream flag set using the SpiffeID CRD. This project is missing support for this.

### PR DESCRIPTION
**Pull request by keeganwitt**
I looked at all the places using `Admin` and used that to determine what I probably needed for adding `Downstream`. I haven't tested that this actually works though.

### CODE REVIEW

**Comment by keeganwitt** [comment]
[This code](https://github.com/spiffe/spire-controller-manager/blob/44d149432422999895335acc36b51255c3865d61/pkg/spireentry/entries.go#L39) seems to only support a `k8s_psat` parent, but if you are using another method (for example `x509pop` to have secondary SPIRE servers in different K8s clusters), then there's no way to specify the parent in the CRD.

```go
parentID, err := spiffeid.FromPathf(trustDomain, "/spire/agent/k8s_psat/%s/%s", clusterName, node.UID)
```

**Comment by keeganwitt** [comment]
> [This code](https://github.com/spiffe/spire-controller-manager/blob/44d149432422999895335acc36b51255c3865d61/pkg/spireentry/entries.go#L39) seems to only support a `k8s_psat` parent, but if you are using another method (for example `x509pop` to have secondary SPIRE servers in different K8s clusters), then there's no way to specify the parent in the CRD.
> 
> ```go
> parentID, err := spiffeid.FromPathf(trustDomain, "/spire/agent/k8s_psat/%s/%s", clusterName, node.UID)
> ```
> 
> I think this would need to be changed in this PR also.

Actually, this might have just been a poorly chosen name ID. I think this can just be updated

**Comment by faisal-memon** [comment]
Do we need to bump the version of the crd for this change?

**Comment by faisal-memon** [comment]
> [This code](https://github.com/spiffe/spire-controller-manager/blob/44d149432422999895335acc36b51255c3865d61/pkg/spireentry/entries.go#L39) seems to only support a `k8s_psat` parent, but if you are using another method (for example `x509pop` to have secondary SPIRE servers in different K8s clusters), then there's no way to specify the parent in the CRD.
> 
> ```go
> parentID, err := spiffeid.FromPathf(trustDomain, "/spire/agent/k8s_psat/%s/%s", clusterName, node.UID)
> ```
> 
> I think this would need to be changed in this PR also.

Were you able to specify the parent with k8s-workload-registrar?

**Comment by keeganwitt** [comment]
> > [This code](https://github.com/spiffe/spire-controller-manager/blob/44d149432422999895335acc36b51255c3865d61/pkg/spireentry/entries.go#L39) seems to only support a `k8s_psat` parent, but if you are using another method (for example `x509pop` to have secondary SPIRE servers in different K8s clusters), then there's no way to specify the parent in the CRD.
> > ```go
> > parentID, err := spiffeid.FromPathf(trustDomain, "/spire/agent/k8s_psat/%s/%s", clusterName, node.UID)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I think this would need to be changed in this PR also.
> 
> Were you able to specify the parent with k8s-workload-registrar?

Yes, this is what we're doing today with the Registrar.

**Comment by keeganwitt** [comment]
> Do we need to bump the version of the crd for this change?

Yea, I suppose so. Bumped.

File: pkg/spireentry/entries.go [hunk 1]
```diff
@@ -37,6 +37,9 @@ func renderPodEntry(spec *spirev1alpha1.ParsedClusterSPIFFEIDSpec, node *corev1.
 		{Type: "k8s", Value: fmt.Sprintf("pod-uid:%s", pod.UID)},
 	}
 	parentID, err := spiffeid.FromPathf(trustDomain, "/spire/agent/k8s_psat/%s/%s", clusterName, node.UID)
+	if spec.ParentID.String() != "" {
```
Comment (keeganwitt) [inline pkg/spireentry/entries.go:40]: The idea here was to use the parent ID from the entry if it exists, otherwise use the generated one above.

**Comment by faisal-memon** [comment]
> > Do we need to bump the version of the crd for this change?
> 
> Yea, I suppose so. Bumped.

I think I was mistaken, we don't need to bump the version for just adding a new field. https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#adding-a-field

File: api/v1alpha1/clusterspiffeid_types.go [hunk 2]
```diff
@@ -67,6 +67,12 @@ type ClusterSPIFFEIDSpec struct {
 	// administrative APIs. Extra care should be taken to only apply this
 	// SPIFFE ID to admin workloads.
 	Admin bool `json:"admin,omitempty"`
+
+	// Downstream indicates that the entry describes a downstream SPIRE server.
+	Downstream bool `json:"downstream,omitempty"`
+
+	// ParentID is the SPIFFE ID of the parent entry.
+	ParentID string `json:"parentId,omitempty"`
```
Comment (faisal-memon) [inline api/v1alpha1/clusterspiffeid_types.go:75]: Can we decouple the custom parent id from downstream and focus on just downstream with this pr? Parenting requires a little more thought.

**Comment by keeganwitt** [comment]
Created !54 for the parentID part.

File: api/v1alpha1/clusterspiffeid_types.go (hunk 2 above)
Comment (keeganwitt) [reply api/v1alpha1/clusterspiffeid_types.go]: OK. Dropped that commit.

**Comment by keeganwitt** [comment]
> > > Do we need to bump the version of the crd for this change?
> > 
> > 
> > Yea, I suppose so. Bumped.
> 
> I think I was mistaken, we don't need to bump the version for just adding a new field. https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#adding-a-field

OK. Dropped that commit.

**Comment by faisal-memon** [comment]
Thanks @keeganwitt for the updates. Changes look good to me.

**Review by azdagron — COMMENTED** [review]
Thanks for opening this contribution, @keeganwitt!

Looks like a spot was missed in api/v1alpha1/clusterspiffeid_webhook.go where the spec is parsed into the ParsedClusterSPIFFEID struct. Downstream is not being set on the parsed spec, which ultimately results in the downstream bit not being set on the registration entry.

I'd encourage you to test out your changes before submitting. It would have found the issue.

**Comment by keeganwitt** [comment]
> Thanks for opening this contribution, @keeganwitt!
> 
> Looks like a spot was missed in api/v1alpha1/clusterspiffeid_webhook.go where the spec is parsed into the ParsedClusterSPIFFEID struct. Downstream is not being set on the parsed spec, which ultimately results in the downstream bit not being set on the registration entry.
> 
> I'd encourage you to test out your changes before submitting. It would have found the issue.

No testing, as noted in the description, this was my best guess what changes were needed. Do you have a recommendation for how to test? I couldn't test with my deployment because we're still using the registrar. Sounds like we could use a CONTRIBUTING.md?

File: config/crd/bases/spire.spiffe.io_clusterspiffeids.yaml [hunk 3]
```diff
@@ -48,6 +48,9 @@ spec:
                 items:
                   type: string
                 type: array
+              downstream:
+                description: Downstream indicates that the entry describes a downstream SPIRE server.
```
[the remark this answers is not in the document]
Comment (keeganwitt) [reply config/crd/bases/spire.spiffe.io_clusterspiffeids.yaml:52]: No. I didn't know about that script. I did this by hand. We should document this in a CONTRIBUTING.md.

**Comment by azdagron** [comment]
We definitely need a CONTRIBUTING.md :)

To test, I normally run `make docker-build` and then make some changes to the integration test scripts to exercise things. See demo/test.sh.

File: config/crd/bases/spire.spiffe.io_clusterspiffeids.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (keeganwitt) [reply config/crd/bases/spire.spiffe.io_clusterspiffeids.yaml:52]: I ran this script now and don't see any differences on my machine 🤷 I'm on Windows though, so I had to run in Docker.

```
docker run --rm -v %cd%:/project -w /project golang:1.19 make manifests
```

**Review by azdagron — APPROVED** [review]
Thanks, @keeganwitt. Looks good.

**Pull request merged**